cmd/kube-apiserver/app/server.go

```diff
@@ -145,6 +145,25 @@ func Run(s *options.APIServer) error {
 		glog.Fatalf("Failed to start kubelet client: %v", err)
 	}
 
+	if s.StorageConfig.DeserializationCacheSize == 0 {
+		// When size of cache is not explicitly set, estimate its size based on
+		// target memory usage.
+		glog.V(2).Infof("Initalizing deserialization cache size based on %dMB limit", s.TargetRAMMB)
+
+		// This is the heuristics that from memory capacity is trying to infer
+		// the maximum number of nodes in the cluster and set cache sizes based
+		// on that value.
+		// From our documentation, we officially recomment 120GB machines for
+		// 2000 nodes, and we scale from that point. Thus we assume ~60MB of
+		// capacity per node.
```

Comment:
our max object size is 900kb, IIRC. I would suggest an alternate heuristic, which is to decide e.g. 1/3 of this memory will be used for the deserialization cache, and then divide by the max object size to see how big the cache should be.

(It would of course be better to just actually measure the collective sizes of the objects in the cache but that's much much harder. A TODO mentioning this might be good.)